src/SimpleGraphs/generators/staticgraphs.jl

```diff
@@ -536,6 +536,60 @@ function double_binary_tree(k::Integer)
     return g
 end
 
+"""
+    regular_tree([T=Int64, ], k::Integer, z::integer)
+
+Create a regular tree or [perfect z-ary tree](https://en.wikipedia.org/wiki/M-ary_tree#Types_of_m-ary_trees): 
+a `k`-level tree where all nodes except the leaves have exactly `z` children. 
+For `z = 2` one recovers a binary tree.
+
+# Examples
+```jldoctest
+julia> regular_tree(4, 3)
+{40, 39} undirected simple Int64 graph
+
+julia> regular_tree(Int8, 3, 2)
+{7, 6} undirected simple Int8 graph
+
+julia> regular_tree(5, 2) == binary_tree(5)
+true
+```
+"""
+function regular_tree(T::Type{<:Integer}, k::Integer, z::Integer)
+    z <= 0 && throw(DomainError(z, "number of children must be positive"))
+    z == 1 && return path_graph(T(k))
+    k <= 0 && return SimpleGraph(zero(T))
+    k == 1 && return SimpleGraph(one(T))
+    if Graphs.isbounded(k) && (BigInt(z)^k - 1) ÷ (z - 1) > typemax(T)
+        throw(InexactError(:convert, T, (BigInt(z)^k - 1) ÷ (z - 1)))
+    end
+
+    n = T((z^k - 1) / (z - 1))
```

Comment:
You already calculate the number of vertices with `(BigInt(z)^k - 1) ÷ (z - 1)`, so we we could just reuse this for `n`.

With the expression `n = T((z^k - 1) / (z - 1))` there are two issues:
- `z^k` might overflow before converting it to `T`
- By using the `/` operator instead of `÷` your intermediate result will be a double

Reply:
A lot of these overflow issues probably won't happen if we restrict  `k` and `z` to `Int` in the function definition.